clients/src/main/java/org/apache/kafka/common/network/SslTransportLayer.java

```diff
@@ -903,6 +901,54 @@ private void updateBytesBuffered(boolean madeProgress) {
 
     @Override
     public long transferFrom(FileChannel fileChannel, long position, long count) throws IOException {
-        return fileChannel.transferTo(position, count, this);
+        if (state == State.CLOSING)
+            throw closingException();
+        if (state != State.READY)
+            return 0;
+
+        if (!flush(netWriteBuffer))
+            return 0;
+
+        long channelSize = fileChannel.size();
+        if (position > channelSize)
+            return 0;
+        int bytesToWrite = (int) Math.min(Math.min(count, channelSize - position), Integer.MAX_VALUE);
+
+        if (fileChannelBuffer == null) {
+            // same size as `netWriteBuffer`
+            int transferSize = 16384;
+            // Allocate a direct buffer to avoid one heap to heap buffer copy. SSLEngine copies the source
```

Comment:
Did we save a heap to heap copy? I thought we only saved the reallocation of new buffers.

Reply:
I discussed with @junrao offline. We save a copy because by passing a direct buffer to `FileChannel.read` we avoid a copy in the standard library code from direct to heap buffer.